File: src/exoticatechnologies/modifications/exotics/impl/PlasmaFluxCatalyst.java
package exoticatechnologies.modifications.exotics.impl;

import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.ui.TooltipMakerAPI;
import com.fs.starfarer.api.ui.UIComponentAPI;
import exoticatechnologies.modifications.exotics.Exotic;
import exoticatechnologies.modifications.ShipModifications;
import exoticatechnologies.util.StringUtils;
import exoticatechnologies.util.Utilities;
import lombok.Getter;

import java.awt.*;
import java.util.HashMap;
import java.util.Map;

public class PlasmaFluxCatalyst extends Exotic {
    private static final String ITEM = "et_plasmacatalyst";

    private static final Map<ShipAPI.HullSize, Integer> MAX_FLUX_EQUIPMENT = new HashMap<>();

    static {
        MAX_FLUX_EQUIPMENT.put(ShipAPI.HullSize.FIGHTER, 10);
        MAX_FLUX_EQUIPMENT.put(ShipAPI.HullSize.FRIGATE, 10);
        MAX_FLUX_EQUIPMENT.put(ShipAPI.HullSize.DESTROYER, 20);
        MAX_FLUX_EQUIPMENT.put(ShipAPI.HullSize.CRUISER, 30);
        MAX_FLUX_EQUIPMENT.put(ShipAPI.HullSize.CAPITAL_SHIP, 50);
    }

    @Getter private final Color mainColor = new Color(0x00BBFF);

    @Override
    public boolean canAfford(CampaignFleetAPI fleet, MarketAPI market) {
        return Utilities.hasItem(fleet.getCargo(), ITEM);
    }

    @Override
    public boolean removeItemsFromFleet(CampaignFleetAPI fleet, FleetMemberAPI fm) {
        Utilities.takeItemQuantity(fleet.getCargo(), ITEM, 1);

        return true;
    }

    @Override
    public Map<String, Float> getResourceCostMap(FleetMemberAPI fm, ShipModifications mods, MarketAPI market) {
        Map<String, Float> resourceCosts = new HashMap<>();
        resourceCosts.put(Utilities.formatSpecialItem(ITEM), 1f);
        return resourceCosts;
    }

    @Override
    public void modifyToolTip(TooltipMakerAPI tooltip, UIComponentAPI title, FleetMemberAPI fm, ShipModifications systems, boolean expand) {
        if (expand) {
            int maxCaps = (int) fm.getFleetCommander().getStats().getMaxCapacitorsBonus().computeEffective(MAX_FLUX_EQUIPMENT.get(fm.getHullSpec().getHullSize()));
            int maxVents = (int) fm.getFleetCommander().getStats().getMaxVentsBonus().computeEffective(MAX_FLUX_EQUIPMENT.get(fm.getHullSpec().getHullSize()));

            StringUtils.getTranslation(this.getKey(), "longDescription")
                    .format("capacitorLimit", Math.ceil(maxCaps / 3f))
                    .format("ventLimit", Math.ceil(maxVents / 3f))
                    .format("crDecrease", 1)
                    .addToTooltip(tooltip, title);
        }
    }

    @Override
    public void applyExoticToStats(FleetMemberAPI fm, MutableShipStatsAPI stats, float bandwidth, String id) {
        if (fm.getFleetCommander() == null) {
            return;
        }

        if (isNPC(fm)) {
            return;
        }

        int numCapsStats = stats.getVariant().getNumFluxCapacitors();
        int numVentsStats = stats.getVariant().getNumFluxVents();

        int maxCaps = (int) fm.getFleetCommander().getStats().getMaxCapacitorsBonus().computeEffective(MAX_FLUX_EQUIPMENT.get(fm.getHullSpec().getHullSize()));
        int maxVents = (int) fm.getFleetCommander().getStats().getMaxVentsBonus().computeEffective(MAX_FLUX_EQUIPMENT.get(fm.getHullSpec().getHullSize()));

        int crReduction = 0;
        if(numCapsStats > Math.ceil(maxCaps / 3f)) {
            crReduction += numCapsStats - Math.ceil(maxCaps / 3f);
        }

        if(numVentsStats > Math.ceil(maxVents / 3f)) {
            crReduction += numVentsStats - Math.ceil(maxVents / 3f);
        }

        if(crReduction > 0) {
            stats.getMaxCombatReadiness().modifyFlat(this.getName(), -crReduction / 100f, this.getName());
        }
    }

    @Override
    public void applyExoticToShip(FleetMemberAPI fm, ShipAPI ship, float bandwidth, String id) {
        int numCaps = ship.getVariant().getNumFluxCapacitors();
        int numVents = ship.getVariant().getNumFluxVents();

        ship.getMutableStats().getFluxCapacity().modifyFlat(this.getBuffId(), numCaps * 200);
        ship.getMutableStats().getFluxDissipation().modifyFlat(this.getBuffId(), numVents * 10);
    }
}

File: src/exoticatechnologies/modifications/exotics/impl/PlasmaFluxCatalyst.kt
package exoticatechnologies.modifications.exotics.impl

import com.fs.starfarer.api.campaign.CampaignFleetAPI
import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.combat.ShipAPI.HullSize
import com.fs.starfarer.api.fleet.FleetMemberAPI
import com.fs.starfarer.api.ui.TooltipMakerAPI
import com.fs.starfarer.api.ui.UIComponentAPI
import exoticatechnologies.modifications.ShipModifications
import exoticatechnologies.modifications.exotics.Exotic
import exoticatechnologies.modifications.exotics.ExoticData
import exoticatechnologies.util.FleetMemberUtils
import exoticatechnologies.util.StringUtils
import exoticatechnologies.util.Utilities
import org.json.JSONObject
import java.awt.Color
import java.util.*
import kotlin.math.ceil

class PlasmaFluxCatalyst(key: String, settings: JSONObject) : Exotic(key, settings) {
    override var color = Color(0x00BBFF)

    override fun canAfford(fleet: CampaignFleetAPI, market: MarketAPI?): Boolean {
        return Utilities.hasItem(fleet.cargo, ITEM)
    }

    override fun removeItemsFromFleet(fleet: CampaignFleetAPI, member: FleetMemberAPI, market: MarketAPI?): Boolean {
        Utilities.takeItemQuantity(fleet.cargo, ITEM, 1f)
        return true
    }

    override fun modifyToolTip(
        tooltip: TooltipMakerAPI,
        title: UIComponentAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData,
        expand: Boolean
    ) {
        if (expand) {
            val maxCaps = getMaxCaps(member, mods, exoticData)
            val maxVents = getMaxVents(member, mods, exoticData)
            StringUtils.getTranslation(key, "longDescription")
                .format("effectLevel", getPositiveMult(member, mods, exoticData) * 100f)
                .format("capacitorLimit", ceil((maxCaps / 3f).toDouble()))
                .format("ventLimit", ceil((maxVents / 3f).toDouble()))
                .format("crDecrease", 1)
                .addToTooltip(tooltip, title)
        }
    }

    fun getMaxCaps(member: FleetMemberAPI, mods: ShipModifications, exoticData: ExoticData): Int {
        return ((FleetMemberUtils.getFleetCommander(member)?.stats?.maxCapacitorsBonus?.computeEffective(MAX_FLUX_EQUIPMENT[member.hullSpec.hullSize]!!.toFloat())
            ?: MAX_FLUX_EQUIPMENT[member.hullSpec.hullSize]!!)
            .toFloat() / getNegativeMult(member, mods, exoticData)).toInt()
    }

    fun getMaxVents(member: FleetMemberAPI, mods: ShipModifications, exoticData: ExoticData): Int {
        return ((FleetMemberUtils.getFleetCommander(member)?.stats?.maxVentsBonus?.computeEffective(MAX_FLUX_EQUIPMENT[member.hullSpec.hullSize]!!.toFloat())
        ?: MAX_FLUX_EQUIPMENT[member.hullSpec.hullSize]!!)
        .toFloat() / getNegativeMult(member, mods, exoticData)).toInt()
    }

    override fun applyExoticToStats(
        id: String,
        stats: MutableShipStatsAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
        if (FleetMemberUtils.getFleetCommander(member) == null) {
            return
        }
        if (isNPC(member)) {
            return
        }

        val numCapsStats = stats.variant.numFluxCapacitors
        val numVentsStats = stats.variant.numFluxVents
        val maxCaps = getMaxCaps(member, mods, exoticData)
        val maxVents = getMaxVents(member, mods, exoticData)

        var crReduction = 0
        if (numCapsStats > ceil((maxCaps / 3f).toDouble())) {
            crReduction += (numCapsStats - ceil((maxCaps / 3f).toDouble())).toInt()
        }
        if (numVentsStats > ceil((maxVents / 3f).toDouble())) {
            crReduction += (numVentsStats - ceil((maxVents / 3f).toDouble())).toInt()
        }
        if (crReduction > 0) {
            stats.maxCombatReadiness.modifyFlat(name, -crReduction / 100f, name)
        }
    }

    override fun applyToShip(
        id: String,
        member: FleetMemberAPI,
        ship: ShipAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
        val numCaps = ship.variant.numFluxCapacitors
        val numVents = ship.variant.numFluxVents
        ship.mutableStats.fluxCapacity.modifyFlat(buffId,numCaps * 200 * getPositiveMult(member, mods, exoticData))
        ship.mutableStats.fluxDissipation.modifyFlat(buffId, numVents * 10 * getPositiveMult(member, mods, exoticData))
    }

    companion object {
        private const val ITEM = "et_plasmacatalyst"
        private val MAX_FLUX_EQUIPMENT: MutableMap<HullSize, Int> = EnumMap(HullSize::class.java)

        init {
            MAX_FLUX_EQUIPMENT[HullSize.FIGHTER] = 10
            MAX_FLUX_EQUIPMENT[HullSize.FRIGATE] = 10
            MAX_FLUX_EQUIPMENT[HullSize.DESTROYER] = 20
            MAX_FLUX_EQUIPMENT[HullSize.CRUISER] = 30
            MAX_FLUX_EQUIPMENT[HullSize.CAPITAL_SHIP] = 50
        }
    }
}